tsl/src/bgw_policy/policy_utils.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <utils/builtins.h>
+#include "dimension.h"
+#include "jsonb_utils.h"
+#include "policy_utils.h"
+
+/* Helper function to compare jsonb label value
+ * and passed in value.
+ * This function is used for labels
+ * defined on the hypertable's dimension
+ * Parameters:
+ * config - jsonb config value
+ * label - label we are looking for inside the config
+ * partitioning_type - Oid for hypertable's dimension column
+ * lag_value - value we will compare against the config's
+ *             value for the label
+ * lag_type - Oid for lag_value
+ * Returns:
+ *    True, if config value is equal to lag_value
+ */
+bool
+policy_config_check_ht_lag_equality(Jsonb *config, const char *json_label, Oid partitioning_type,
+									Oid lag_type, Datum lag_datum)
+{
+	if (IS_INTEGER_TYPE(partitioning_type))
+	{
+		bool found;
+		int64 config_value = ts_jsonb_get_int64_field(config, json_label, &found);
+
+		if (!found)
+			ereport(ERROR,
+					(errcode(ERRCODE_INTERNAL_ERROR),
+					 errmsg("could not find %s in config for job", json_label)));
+
+		switch (lag_type)
+		{
+			case INT2OID:
+				return config_value == DatumGetInt16(lag_datum);
+			case INT4OID:
+				return config_value == DatumGetInt32(lag_datum);
+			case INT8OID:
+				return config_value == DatumGetInt64(lag_datum);
+			default:
+				return false;
+		}
+	}
+	else
+	{
+		if (lag_type != INTERVALOID)
+			return false;
+		Interval *config_value = ts_jsonb_get_interval_field(config, json_label);
+		if (config_value == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_INTERNAL_ERROR),
+					 errmsg("could not find %s in config for job", json_label)));
+
+		return DatumGetBool(
+			DirectFunctionCall2(interval_eq, IntervalPGetDatum(config_value), lag_datum));
+	}
+}
+
+Datum
+subtract_integer_from_now(int64 interval, Oid time_dim_type, Oid now_func)
+{
+	Datum now;
+	int64 res;
+
+	AssertArg(IS_INTEGER_TYPE(time_dim_type));
+
+	now = OidFunctionCall0(now_func);
+
+	switch (time_dim_type)
+	{
+		case INT2OID:
+			res = DatumGetInt16(now) - interval;
+			if (res < PG_INT16_MIN || res > PG_INT16_MAX)
+				ereport(ERROR,
+						(errcode(ERRCODE_INTERVAL_FIELD_OVERFLOW), errmsg("ts_interval overflow")));
+			return Int16GetDatum(res);
+		case INT4OID:
+			res = DatumGetInt32(now) - interval;
+			if (res < PG_INT32_MIN || res > PG_INT32_MAX)
+				ereport(ERROR,
+						(errcode(ERRCODE_INTERVAL_FIELD_OVERFLOW), errmsg("ts_interval overflow")));
```

Comment:
The suggested error message is too generic and we will not be able to pin point the problem if this happens. (This would be part of policy execution and logged).